apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/ExplanationOfBenefitResourceProvider.java

```diff
@@ -350,75 +375,113 @@ public Bundle findByPatient(
       return TransformerUtils.createBundle(paging, eobs, loadedFilterManager.getTransactionTime());
     }
 
+    /*
+     * execute a database function that returns a bitwise mask value that denotes that the given
+     * claim type will have data for the specified beneficiayrId. This represents fast and efficient
+     * way to ignore a requested claim type that ultimately has no data for our beneficiaryId.
+     */
+    List<Object[]> values =
+        entityManager
+            .createNativeQuery(CHECK_CLAIMS_FOR_DATA)
+            .setParameter("beneIdValue", beneficiaryId)
+            .getResultList();
+
+    Integer maskVal = (Integer) (values != null && values.size() > 0 ? values.get(0) : 0);
+
+    /*
+     * There is another minor efficiency that here if the database function does not perform an
+     * efficacy check that the bene_id exists. Essentially, if the mask value returned from the
+     * function is zero, we can simply return and empty claims bundle.
+     */
+    /*
+    if (maskVal == 0) {
+      // Add bene_id to MDC logs
+      LoggingUtils.logBeneIdToMdc(beneficiaryId);
+      // Add number of resources to MDC logs
+      LoggingUtils.logResourceCountToMdc(0);
+
+      return TransformerUtils.createBundle(paging, eobs, loadedFilterManager.getTransactionTime());
+    }
+    */
+
     /*
      * The way our JPA/SQL schema is setup, we have to run a separate search for
      * each claim type, then combine the results. It's not super efficient, but it's
      * also not so inefficient that it's worth fixing.
      */
-    if (claimTypes.contains(ClaimType.CARRIER))
+    if (claimTypes.contains(ClaimType.CARRIER) && (maskVal & V_CARRIER_HAS_DATA) == 0) {
```

Comment:
I would suggest parallelizing this logic prior to additional thought into pre-flight check optimizations, as it can potentially offer many more efficiency gains outside of just the query, and most of the pre-flight check gains could potentially be due to the simple serial aspect of the current implementation.